## 1. 2331 반복수열

#### 문제
다음과 같이 정의된 수열이 있다.  
  
D[1] = A  
D[n] = D[n-1]의 각 자리의 숫자를 P번 곱한 수들의 합  
예를 들어 A=57, P=2일 때, 수열 D는 {57, 74(=5^2+7^2=25+49), 65, 61, 37, 58, 89, 145, 42, 20, 4, 16, 37, …}이 된다.  
그 뒤에는 앞서 나온 수들(57부터가 아니라 58부터)이 반복된다.  
  
이와 같은 수열을 계속 구하다 보면 언젠가 이와 같은 반복수열이 된다. 이때, 반복되는 부분을 제외했을 때,  
수열에 남게 되는 수들의 개수를 구하는 프로그램을 작성하시오. 위의 예에서는 {57, 74, 65, 61}의 네 개의 수가 남게 된다.  
  
#### 입력
첫째 줄에 A(1 ≤ A ≤ 9999), P(1 ≤ P ≤ 5)가 주어진다.  
  
#### 출력
첫째 줄에 반복되는 부분을 제외했을 때, 수열에 남게 되는 수들의 개수를 출력한다.  

### ------------------------------------------------------------------------------

계산하다가 중복되면 시작점 이전까지 길이만..

In [2]:
# 60ms
A, P = input().split()
P = int(P)
sq = [A]
nxt = str(sum([int(i)**P for i in A]))
while nxt not in sq:
    sq.append(nxt)
    nxt = str(sum([int(i)**P for i in nxt]))
print(sq.index(nxt))

57 2
4


In [5]:
# 56ms  set으로 하면 in 연산이 O(1) 리스트로하면 O(N)
A, P = input().split()
P = int(P)
sq = [A]
sq_s = set(sq)
nxt = str(sum([int(i)**P for i in A]))
while nxt not in sq_s:
    sq.append(nxt)
    sq_s.add(nxt)
    nxt = str(sum([int(i)**P for i in nxt]))
print(sq.index(nxt))

57 2
4


In [ ]:
# 백준..
a,p=input().split()
d={}
i=0
while(a in d)<1:        # not a in d를 이렇게? 왜?? 1byte 짧아서??????
    d[a]=i
    i+=1
    a=str(sum(int(c)**int(p)for c in a))
print(d[a])

In [13]:
import time
a = [i for i in range(20000)]
b = set(a)
t1 = time.clock()
for i in range(20000):
    if i in a:
        pass
t2 = time.clock()
print(t2-t1)

t3 = time.clock()
for i in range(20000):
    if i in b:
        pass
t4 = time.clock()
print(t4-t3)

12.922678454862307
0.009568842706812575


##### 2. 9466 텀 프로젝트

#### 문제
이번 가을학기에 '문제 해결' 강의를 신청한 학생들은 텀 프로젝트를 수행해야 한다. 프로젝트 팀원 수에는 제한이 없다.  
심지어 모든 학생들이 동일한 팀의 팀원인 경우와 같이 한 팀만 있을 수도 있다.  
프로젝트 팀을 구성하기 위해, 모든 학생들은 프로젝트를 함께하고 싶은 학생을 선택해야 한다. (단, 단 한 명만 선택할 수 있다.)  
혼자 하고 싶어하는 학생은 자기 자신을 선택하는 것도 가능하다.  
  
학생들이($s_1, s_2, ..., s_r$)이라 할 때, r=1이고 $s_1$이 $s_1$을 선택하는 경우나, $s_1$이 $s_2$를 선택하고, $s_2$가 $s_3$를 선택하고,...,  
$s_{r-1}$이 $s_r$을 선택하고, $s_r$이 $s_1$을 선택하는 경우에만 한 팀이 될 수 있다.  
  
예를 들어, 한 반에 7명의 학생이 있다고 하자. 학생들을 1번부터 7번으로 표현할 때, 선택의 결과는 다음과 같다.  
  
| 1 | 2 | 3 | 4 | 5 | 6 | 7 |
|---|---|---|---|---|---|---|
| 3 | 1 | 3 | 7 | 3 | 4 | 6 |  
위의 결과를 통해 (3)과 (4, 7, 6)이 팀을 이룰 수 있다. 1, 2, 5는 어느 팀에도 속하지 않는다.  
  
주어진 선택의 결과를 보고 어느 프로젝트 팀에도 속하지 않는 학생들의 수를 계산하는 프로그램을 작성하라.  
  
#### 입력
첫째 줄에 테스트 케이스의 개수 T가 주어진다. 각 테스트 케이스의 첫 줄에는 학생의 수가 정수 n (2 ≤ n ≤ 100,000)으로 주어진다.  
각 테스트 케이스의 둘째 줄에는 선택된 학생들의 번호가 주어진다. (모든 학생들은 1부터 n까지 번호가 부여된다.)  
  
#### 출력
각 테스트 케이스마다 한 줄에 출력하고, 각 줄에는 프로젝트 팀에 속하지 못한 학생들의 수를 나타내면 된다.  

### ------------------------------------------------------------------------------

## Markdown Table
http://www.tablesgenerator.com/markdown_tables  
  
| Stretch/Untouched | ProbDistribution | Accuracy |  
| --- | --- | --- |  
| Stretched | Gaussian | .843 |  
밑에 처럼 하는데, 한줄 띄고해야  
| Stretch/Untouched | ProbDistribution | Accuracy |  
| --- | --- | --- |  
| Stretched | Gaussian | .843 |  

Cycle이 존재하면 빼주기..자기 자신으로 가도 빼주기..없어진 애한테 향하면 걘 혼자...

In [12]:
# import sys
# input = sys.stdin.readline
T = int(input())
for tc in range(T):
    n = int(input())
    s = [0] + list(map(int, input().split()))    # 고른 정보....
    nodes = {i for i in range(1, n+1)}
    no_team = 0
    print('-'*20)
    while nodes:
        nxt = nodes.pop()
        print(f'start node: {nxt}')
        d={}
        idx = 0
        d[nxt] = 0
        cur = set([nxt])
        nxt = s[nxt]
        print(f'next visit node: {nxt}')
        # cycle여부에 따라 더해줘야 할 값이 다름.. 
        while nxt in nodes:      # 다음 방문할 노드가 없어졌다면, (cycle이 생기거나, 이미 팀이 됐거나.. 검사해서 혼자가 됐거나..)
            nodes.remove(nxt)        
            idx += 1
            d[nxt] = idx
            cur.add(nxt)
            nxt = s[nxt]        
            print(f'next visit node: {nxt}')
        print(f'방문 경로: {cur}')
        if nxt in cur:                    # cycle
            no_team += d[nxt]
        else:                             # cycle x
            no_team += idx + 1
        print(f'current answer = {no_team}')
        print('-'*20)
    print(no_team)

2
7
3 1 3 7 3 4 6
--------------------
start node: 1
next visit node: 3
next visit node: 3
방문 경로: {1, 3}
current answer = 1
--------------------
start node: 2
next visit node: 1
방문 경로: {2}
current answer = 2
--------------------
start node: 4
next visit node: 7
next visit node: 6
next visit node: 4
방문 경로: {4, 6, 7}
current answer = 2
--------------------
start node: 5
next visit node: 3
방문 경로: {5}
current answer = 3
--------------------
3
8
1 2 3 4 5 6 7 8
--------------------
start node: 1
next visit node: 1
방문 경로: {1}
current answer = 0
--------------------
start node: 2
next visit node: 2
방문 경로: {2}
current answer = 0
--------------------
start node: 3
next visit node: 3
방문 경로: {3}
current answer = 0
--------------------
start node: 4
next visit node: 4
방문 경로: {4}
current answer = 0
--------------------
start node: 5
next visit node: 5
방문 경로: {5}
current answer = 0
--------------------
start node: 6
next visit node: 6
방문 경로: {6}
current answer = 0
--------------------
start node: 7


너무 느리다...5608ms. 근데 1등도 2324ms네..

In [ ]:
# 백준 파이썬 1등
for _ in range(int(input())):
    n = int(input())
    ind, ck = [0] * (n+1), [0] * (n+1)
    h,r = 0,0
    nxt = [0] + list(map(int,input().split()))
    for i in nxt:
        ind[i] += 1
    for i in range(1,n+1):
        h = i
        while ind[h] == 0 and ck[h] == 0:
            ck[h] = 1
            ind[nxt[h]] -= 1
            h = nxt[h]
            r += 1
    print(r)

2  
7  
3 1 3 7 3 4 6  
8  
1 2 3 4 5 6 7 8  
 
 ---
3  
0

## 3. 2667 단지 번호 붙이기

#### 문제
<그림 1>과 같이 정사각형 모양의 지도가 있다. 1은 집이 있는 곳을, 0은 집이 없는 곳을 나타낸다.  
철수는 이 지도를 가지고 연결된 집들의 모임인 단지를 정의하고, 단지에 번호를 붙이려 한다.  
여기서 연결되었다는 것은 어떤 집이 좌우, 혹은 아래위로 다른 집이 있는 경우를 말한다. 대각선상에 집이 있는 경우는 연결된 것이 아니다.  
<그림 2>는 <그림 1>을 단지별로 번호를 붙인 것이다.  
지도를 입력하여 단지수를 출력하고, 각 단지에 속하는 집의 수를 오름차순으로 정렬하여 출력하는 프로그램을 작성하시오.  
<img src="https://www.acmicpc.net/upload/images/ITVH9w1Gf6eCRdThfkegBUSOKd.png">


#### 입력
첫 번째 줄에는 지도의 크기 N(정사각형이므로 가로와 세로의 크기는 같으며 5≤N≤25)이 입력되고,  
그 다음 N줄에는 각각 N개의 자료(0혹은 1)가 입력된다.  
  
#### 출력
첫 번째 줄에는 총 단지수를 출력하시오. 그리고 각 단지내 집의 수를 오름차순으로 정렬하여 한 줄에 하나씩 출력하시오.  

### ------------------------------------------------------------------------------

BFS로 단지 찾고, 리스트에 단지를 이루는 집의 수 append. 정렬해서 len(list), 한줄씩 오름차순으로 출력.

In [39]:
# import sys
# input = sys.stdin.readline
from collections import deque
N = int(input())
m = [input() for _ in range(N)]     # 지도정보 입력.
v = [[False]*N for _ in range(N)]           # 방문여부 체크.
# 상하좌우 4방향 이동할거..
dx = 1, -1, 0, 0
dy = 0, 0, 1, -1
# 단지정보 저장할 list
c = []
# 전체 돌면서 방문 x이고 0 아니면 단지검사.
for j in range(N):
    for i in range(N):
        if m[j][i] != '0' and not v[j][i]:
            q = deque()
            num = 1
            q.append((j,i))
            v[j][i] = True
            # 상하좌우 4방향 탐색..
            print('-'*20)
            print('단지 찾기 시작')
            while q:
                # 레벨별로 진행
                for _ in range(len(q)):
                    y, x = q.popleft()
                    print(f'현재 좌표: {y}, {x}')
                    for d in range(4):
                        cx = x + dx[d]
                        cy = y + dy[d]            
                        print(f'검사 좌표: {cy}, {cx}')
                        if cx>=0 and cx<N and cy>=0 and cy<N and m[cy][cx]=='1' and not v[cy][cx]:
                            print('pass')
                            q.append((cy,cx))
                            v[cy][cx] = True
                            print(f'큐에 추가, {cy}, {cx}')
                            num += 1
            c.append(num)
c.sort()
print(len(c))
print('\n'.join(map(str, c)))        

7
0110100
0110101
1110101
0000111
0100000
0111110
0111000
--------------------
단지 찾기 시작
현재 좌표: 0, 1
검사 좌표: 0, 2
pass
큐에 추가, 0, 2
검사 좌표: 0, 0
검사 좌표: 1, 1
pass
큐에 추가, 1, 1
검사 좌표: -1, 1
현재 좌표: 0, 2
검사 좌표: 0, 3
검사 좌표: 0, 1
검사 좌표: 1, 2
pass
큐에 추가, 1, 2
검사 좌표: -1, 2
현재 좌표: 1, 1
검사 좌표: 1, 2
검사 좌표: 1, 0
검사 좌표: 2, 1
pass
큐에 추가, 2, 1
검사 좌표: 0, 1
현재 좌표: 1, 2
검사 좌표: 1, 3
검사 좌표: 1, 1
검사 좌표: 2, 2
pass
큐에 추가, 2, 2
검사 좌표: 0, 2
현재 좌표: 2, 1
검사 좌표: 2, 2
검사 좌표: 2, 0
pass
큐에 추가, 2, 0
검사 좌표: 3, 1
검사 좌표: 1, 1
현재 좌표: 2, 2
검사 좌표: 2, 3
검사 좌표: 2, 1
검사 좌표: 3, 2
검사 좌표: 1, 2
현재 좌표: 2, 0
검사 좌표: 2, 1
검사 좌표: 2, -1
검사 좌표: 3, 0
검사 좌표: 1, 0
--------------------
단지 찾기 시작
현재 좌표: 0, 4
검사 좌표: 0, 5
검사 좌표: 0, 3
검사 좌표: 1, 4
pass
큐에 추가, 1, 4
검사 좌표: -1, 4
현재 좌표: 1, 4
검사 좌표: 1, 5
검사 좌표: 1, 3
검사 좌표: 2, 4
pass
큐에 추가, 2, 4
검사 좌표: 0, 4
현재 좌표: 2, 4
검사 좌표: 2, 5
검사 좌표: 2, 3
검사 좌표: 3, 4
pass
큐에 추가, 3, 4
검사 좌표: 1, 4
현재 좌표: 3, 4
검사 좌표: 3, 5
pass
큐에 추가, 3, 5
검사 좌표: 3, 3
검사 좌표: 4, 4
검사 좌표: 2, 4
현재 좌표: 3, 5
검사 좌표: 3, 6
pass
큐에 추가, 3, 6
검사 좌표: 

88ms. 좀 느리다. 왜 dfs보다 느리지...

In [ ]:
# 백준 dfs
def dfs(d, x, y):
    r = 0
    if d[x][y] == '1':
        r = 1
        d[x][y] = '0'
    if 1 <= x and d[x - 1][y] == '1':
        r += dfs(d, x - 1, y)
    if x < len(d[x]) - 1 and d[x + 1][y] == '1':
        r += dfs(d, x + 1, y)
    if 1 <= y and d[x][y - 1] == '1':
        r += dfs(d, x, y - 1)
    if y < len(d) - 1 and d[x][y + 1] == '1':
        r += dfs(d, x, y + 1)
    return r

r, d = [], [list(input()) for i in range(int(input()))]
for i in range(len(d)):
    for j in range(len(d[i])):
        if d[i][j] == '1':
            r.append(dfs(d, i, j))
print(len(r))
[print(x) for x in sorted(r)]

In [ ]:
# 백준 뭔코드냐...
r,d,*a=0,{(x,y)for y in range(int(input()))for x,c in enumerate(input())if'0'<c}
while d:
 s=[d.pop()];r+=1;b=0
 while s:
  x,y=s.pop();b+=1
  for e in-2,0,2,4:
   p=x+e//3,y+e%3-1
   if{p}<=d:d-={p};s+=p,
 a+=b,
print(*[r]+sorted(a))

7  
0110100  
0110101  
1110101  
0000111  
0100000  
0111110  
0111000  

--- 
3  
7  
8  
9

## 4. 4963 섬의 개수

#### 문제
정사각형으로 이루어져 있는 섬과 바다 지도가 주어진다. 섬의 개수를 세는 프로그램을 작성하시오.  
    
<img src="https://www.acmicpc.net/upload/images/island.png">  
    
한 정사각형과 가로, 세로 또는 대각선으로 연결되어 있는 사각형은 걸어갈 수 있는 사각형이다.   
  
두 정사각형이 같은 섬에 있으려면, 한 정사각형에서 다른 정사각형으로 걸어서 갈 수 있는 경로가 있어야 한다.  
지도는 바다로 둘러쌓여 있으며, 지도 밖으로 나갈 수 없다.  
  
#### 입력
입력은 여러 개의 테스트 케이스로 이루어져 있다. 각 테스트 케이스의 첫째 줄에는 지도의 너비 w와 높이 h가 주어진다.  
w와 h는 50보다 작거나 같은 양의 정수이다.  
  
둘째 줄부터 h개 줄에는 지도가 주어진다. 1은 땅, 0은 바다이다.  
  
입력의 마지막 줄에는 0이 두 개 주어진다.  
  
#### 출력
각 테스트 케이스에 대해서, 섬의 개수를 출력한다.  

### ------------------------------------------------------------------------------

대각선도 연결..8방향.

In [44]:
### BFS
# import sys
# input = sys.stdin.readline
from collections import deque
w, h = map(int, input().split())   # 너비 w, 높이 h
# 대각선까지 8방향으로 이동가능.
dx = 1, 1, 1, -1, -1, -1, 0, 0
dy = -1, 0, 1, -1, 0, 1, -1, 1
# 마지막 입력이 주어질 때까지. 
while (w, h) != (0, 0):
    m = [list(map(int, input().split())) for _ in range(h)]  # 지도입력 . 1: 땅 / 0: 바다
    v = [[False] * w for _ in range(h)]                      # 방문여부 체크
    n = 0
    for j in range(h):
        for i in range(w):
            # 방문 안했고, 땅이면..
            if not v[j][i] and m[j][i]:
                v[j][i] = True
                q = deque()
                q.append((j,i))
                while q:
                    for _ in range(len(q)):
                        y, x = q.popleft()
                        for d in range(8):
                            cy = y + dy[d]
                            cx = x + dx[d]
                            if cx>=0 and cx<w and cy>=0 and cy<h and m[cy][cx] and not v[cy][cx]:
                                q.append((cy, cx))
                                v[cy][cx] = True
                n += 1
    print(f'섬의 개수:\n{n}')
    w, h = map(int, input().split())

5 6
1 1 0 1 1
1 0 0 0 1
0 0 1 0 1
0 1 0 0 0
0 0 0 1 1
1 1 0 0 1
섬의 개수:
5
0 0


116ms..